core/input/input.h

```diff
@@ -43,10 +43,16 @@ class Input : public Object {
 
 public:
 	enum MouseMode {
-		MOUSE_MODE_VISIBLE,
-		MOUSE_MODE_HIDDEN,
-		MOUSE_MODE_CAPTURED,
-		MOUSE_MODE_CONFINED
+		MOUSE_MODE_VISIBLE = 0,
+		MOUSE_MODE_HIDDEN = 1,
+		// CAPTURED_VISIBLE is useful internally but not necessary to expose.
+		MOUSE_MODE_CAPTURED_VISIBLE = 2,
+		// Captured should normally imply hidden, so CAPTURED becomes 3.
+		MOUSE_MODE_CAPTURED = MOUSE_MODE_CAPTURED_VISIBLE | MOUSE_MODE_HIDDEN,
+		// Confined and hidden would be 5 (MOUSE_MODE_HIDDEN | MOUSE_MODE_CONFINED).
+		MOUSE_MODE_CONFINED = 4,
+		// This value should never be set, but it's useful for bitwise checking via &.
+		MOUSE_MODE_GRABBED = MOUSE_MODE_CAPTURED_VISIBLE | MOUSE_MODE_CONFINED,
```

Comment:
It's not valid input, because `ERR_FAIL_INDEX((int)p_mode, 6);` excludes `6`.

Reply:
I'm not sure I get why this has to be so complicated. Wouldn't it be easier to just have e.g.
```
MOUSE_MODE_HIDDEN = 1,
MOUSE_MODE_CAPTURED = 2,
MOUSE_MODE_CONFINED = 4,
```
Then check these bits as needed in the code.

I think the MOUSE_MODE_GRABBED and MOUSE_MODE_CAPTURED (your version) just makes things more confusing, not less.

There is no need for MOUSE_MODE_VISIBLE, as by definition if HIDDEN is not set it is visible, and checking bitwise flags for absolute values is bad practice (what happens if you add a new flag?).

And having MOUSE_MODE_CAPTURED_VISIBLE as 2 seems ridiculous because you can then have non-sensical flag combinations, of a HIDDEN and VISIBLE mouse mode...

If on the other hand the idea is to enforce state, you could use e.g. an encapsulated class / struct, either using the above flags or something like:
```
class MouseMode
{
bool visible; // or hidden
enum CaptureState
{
CS_NORMAL,
CS_CAPTURED,
CS_CONFINED,
};
public:
// accessors
// ...
}
```
and use the accessor functions to enforce state.

Another option is combining an enum and flags in the same bits (e.g. use bits 0-3 for an enum, and 4-7 for flags etc). But this is usually to save memory / pass something smaller around, and not really something user-friendly unless you have accessor functions.

That said I'm not really familiar with how MouseMode is used throughout Godot, there may be other considerations.